File: ProblemSolving/Java/TimeInWords.java
//Problem: https://www.hackerrank.com/challenges/the-time-in-words
public class TimeInWords {

    public static void main(String[] args) {
        System.out.println(timeInWords(5, 47)); ////thirteen minutes to six
    }

    public static String timeInWords(int hour, int minute) {
        if (hour < 1 || hour > 12 || minute < 0 || minute > 59)
            return "Incorrect Time";
        if (minute == 0) {
            String hourAsWord = convertNumberToWord(hour);
            return hourAsWord + " o' clock";
        }
        if (minute <= 30) {
            String minuteAsWord = convertNumberToWord(minute);
            String hourAsWord = convertNumberToWord(hour);
            String minuteAsPluralOrSingular = (minute == 1) ? "minute" : "minutes";
            if (minute == 15 || minute == 30) {
                return minuteAsWord + " past " + hourAsWord;
            }
            return minuteAsWord + " " + minuteAsPluralOrSingular + " past " + hourAsWord;
        } else {
            int remainingMinutes = 60 - minute;
            String remainingMinutesAsWord = convertNumberToWord(remainingMinutes);
            String nextHourAsWord = convertNumberToWord(hour + 1);
            String minuteAsPluralOrSingular = (remainingMinutes == 1) ? "minute" : "minutes";
            if (remainingMinutes == 15) {
                return remainingMinutesAsWord + " to " + nextHourAsWord;
            }
            return remainingMinutesAsWord + " " + minuteAsPluralOrSingular + " to " + nextHourAsWord;
        }
    }

    public static String convertNumberToWord(int number) {
        String[] units = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "eleven",
                "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"};
        int unit = (number > 9) ? Character.getNumericValue(Integer.toString(number).charAt(1)) : number;
        switch (number) {
            case 15:
                return "quarter";
            case 20:
                return "twenty";
            case 30:
                return "half";
            default:
                if (number < 20)
                    return units[number];
                else if (number < 30)
                    return "twenty " + units[unit];
                else
                    return "";
        }
    }
}

File: ProblemSolving/Kotlin/TimeInWords.kt
//Problem: https://www.hackerrank.com/challenges/the-time-in-words
fun main(args: Array<String>) {
    println(timeInWords(5, 47)) //thirteen minutes to six
    println(timeInWords(3, 0)) //three o' clock
    println(timeInWords(7, 15)) //quarter past seven
}

fun timeInWords(hour: Int, minute: Int) : String {
    if (hour !in 1..12 || minute !in 0..59) return "Incorrect time"
    if (minute == 0) {
        val hourAsWord = convertNumberToWord(hour)
        return "$hourAsWord o' clock"
    }
    if (minute in 1..30) {
        val minuteAsWord = convertNumberToWord(minute)
        val hourAsWord = convertNumberToWord(hour)
        val minuteAsPluralOrSingular = if (minute == 1) "minute" else "minutes"
        if (minute == 15 || minute == 30) {
            return "$minuteAsWord past $hourAsWord"
        }
        return "$minuteAsWord $minuteAsPluralOrSingular past $hourAsWord"
    } else {
        val remainingMinutes = 60 - minute
        val remainingMinutesAsWord = convertNumberToWord(remainingMinutes)
        val nextHourAsWord = convertNumberToWord(hour + 1)
        val minuteAsPluralOrSingular = if (remainingMinutes == 1) "minute" else "minutes"
        if (remainingMinutes == 15) {
            return "$remainingMinutesAsWord to $nextHourAsWord"
        }
        return "$remainingMinutesAsWord $minuteAsPluralOrSingular to $nextHourAsWord"
    }
}

fun convertNumberToWord(number: Int) : String {
    val units = arrayOf("zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "eleven",
            "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen")
    val unit: Int = if (number > 9) (number % 10) else number
    return when (number) {
        15 -> "quarter"
        in 0..19 -> units.get(number)
        20 -> "twenty"
        in 21..29 -> "twenty " + units.get(unit)
        30 -> "half"
        else -> ""
    }
}